### PR TITLE
Precomputing Transposes for frozen linear layers

### PR DESCRIPTION
**Pull request by Gamrix**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65631

Differential Revision: [D31314248](https://our.internmc.facebook.com/intern/diff/D31314248)

### CODE REVIEW

File: test/jit/test_freezing.py [hunk 1]
```diff
@@ -1,22 +1,20 @@
...
+from torch.testing._internal.jit_utils import JitTestCase
 from torch.utils import mkldnn as mkldnn_utils
 
-
-from torch.jit._recursive import wrap_cpp_module
-from typing import Any
-from itertools import product
-import io
-
 try:
```
Comment (Gamrix) [inline test/jit/test_freezing.py:18]: I guess some linting step I did ran `isort`.

File: torch/csrc/jit/passes/frozen_linear_transpose.cpp [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+      return false;
+    }
+
+    // This also filters out out-variants of the linear op.
+    if (nonConstantParameters(node)) {
+      return false;
+    }
+
+    // Op is only profitable on CUDA
+    auto weight_tensor = constant_as<Tensor>(weight).value();
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_linear_transpose.cpp:40]: I thought our findings were that this was profitable on CPU as well ?

File: torch/csrc/jit/passes/frozen_linear_transpose.cpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  bool run() {
+    handleBlockAndSubblocks(graph_->block());
+    return graph_modified;
+  }
+
+  bool is_constant_linear_op(Node* node) {
+    if (node->kind() != aten::linear) {
+      return false;
+    }
+    auto weight = node->namedInput("weight");
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_linear_transpose.cpp:29]: weight is never none, do you mean bias ? Can we handle the no-bias case as well here ?

Comment (eellison) [reply torch/csrc/jit/passes/frozen_linear_transpose.cpp:29]: looks like bias is actually handled

File: torch/csrc/jit/passes/frozen_linear_transpose.cpp [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+          graph_->create(aten::add, {matmul->output(), bias, bias_scale});
+      bias_result->insertAfter(matmul);
+      node->replaceAllUsesWith(bias_result);
+    }
+    node->destroy();
+  };
+
+  void handleBlockAndSubblocks(Block* block) {
+    // Can't delete nodes while also iterating over it
+    std::vector<Node*> constant_linear_nodes;
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_linear_transpose.cpp:77]: we can iterate like this https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/constant_propagation.cpp#L396

File: torch/csrc/jit/passes/frozen_linear_transpose.cpp [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+      Value* bias_scale = graph_->insertConstant(1);
+      Node* bias_result =
+          graph_->create(aten::add, {matmul->output(), bias, bias_scale});
+      bias_result->insertAfter(matmul);
+      node->replaceAllUsesWith(bias_result);
+    }
+    node->destroy();
+  };
+
+  void handleBlockAndSubblocks(Block* block) {
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_linear_transpose.cpp:75]: this block is a good use case for https://github.com/pytorch/pytorch/blob/3bdee2bbed69c751803339390203362800497619/torch/csrc/jit/runtime/graph_iterator.h#L7

File: torch/csrc/jit/passes/frozen_linear_transpose.h [hunk 6]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include <torch/csrc/jit/ir/ir.h>
+
+namespace torch {
+namespace jit {
+
+// Concats multiple linear ops with the same Tensor input
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_linear_transpose.h:8]: bad copy pasta

File: torch/csrc/jit/passes/frozen_linear_transpose.cpp [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    }
+    for (auto node : constant_linear_nodes) {
+      replace_linear_with_matmul(node);
+    }
+    return;
+  }
+
+ private:
+  std::shared_ptr<Graph> graph_;
+  bool graph_modified = false;
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_linear_transpose.cpp:96]: nit: `graph_modified` -> `graph_modified_`

**Review by eellison — COMMENTED** [review]
Looks good! Can you add more comments to this - both to the PR description detailing the benchmarking that was done and also inline into the pass itself ? 

Also, i thought we determined this was good for CPU as well ?

**Comment by Gamrix** [comment]
Given that we determined in the meeting that single core CPU is the main metric to look at (due to prod uses usually running many concurrent processes with 1 core per process), and that multicore CPU didn't significantly regress, I will remove the restriction on only running this on GPU.

Link to benchmarking results 
https://fb.quip.com/6RmbAN7ZBXrD

File: torch/csrc/jit/passes/frozen_linear_transpose.cpp (hunk 2 above)
Comment (Gamrix) [reply torch/csrc/jit/passes/frozen_linear_transpose.cpp:40]: As 1CPU is the thing  we care about, I have updated it to also optimize for the CPU case.

**Comment by Gamrix** [comment]
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31314248).

**Review by eellison — APPROVED** [review]
